File: solutions/ruby/simple-linked-list/1/simple_linked_list.rb
# Clase que representa cada nodo individual (canción)
class Element
  attr_reader :datum
  attr_accessor :next

  def initialize(datum)
    @datum = datum
    @next = nil
  end
end

# Clase que maneja la estructura de la lista enlazada
class SimpleLinkedList
  attr_reader :head

  def initialize(array = [])
    @head = nil
    # Si nos pasan un arreglo al inicializar, cargamos sus elementos
    # Los recorremos al revés para que el orden de la lista coincida
    array.each { |datum| push(Element.new(datum)) } if array
  end

  # Agrega un elemento al inicio de la lista
  def push(element)
    element.next = @head
    @head = element
    self
  end

  # Remueve y devuelve el elemento al inicio de la lista
  def pop
    return nil if @head.nil?

    popped_element = @head
    @head = @head.next
    popped_element
  end

  # Convierte nuestra lista enlazada en un Arreglo estándar de Ruby
  def to_a
    array = []
    current = @head
    while current
      array << current.datum
      current = current.next
    end
    array
  end

  # Método de clase para instanciar una lista desde un arreglo directamente
  def self.from_a(array)
    return new if array.nil? || array.empty?
    
    # Exercism suele evaluar que la lista mantenga el orden del arreglo original,
    # dado que 'push' inserta al frente, pasamos el arreglo invertido para compensar.
    new(array.reverse)
  end

  # Invierte el orden de los punteros de la lista en su lugar (In-place)
  def reverse!
    previous = nil
    current = @head

    while current
      next_node = current.next  # Guardamos el resto de la lista
      current.next = previous   # Volteamos el puntero actual hacia atrás
      previous = current        # Movemos 'previous' un paso adelante
      current = next_node       # Movemos 'current' un paso adelante
    end

    @head = previous            # El último nodo procesado se convierte en la nueva cabeza
    self
  end
end

File: solutions/java/simple-linked-list/2/src/main/java/SimpleLinkedList.java
import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.List;
import java.util.NoSuchElementException;

class SimpleLinkedList<T> {
    private Node<T> head;
    private int size;

    // Nodo interno para la lista enlazada
    private static class Node<T> {
        T value;
        Node<T> next;

        Node(T value) {
            this.value = value;
        }
    }

    // Constructor vacío
    SimpleLinkedList() {
    }

    // Constructor que recibe un array de valores
    SimpleLinkedList(T[] values) {
        for (T val : values) {
            push(val);
        }
    }

    // Añade un elemento al principio de la lista
    void push(T value) {
        Node<T> newNode = new Node<>(value);
        newNode.next = head;
        head = newNode;
        size++;
    }

    // Elimina y devuelve el elemento del principio de la lista
    T pop() {
        if (head == null) {
            throw new NoSuchElementException();
        }
        T value = head.value;
        head = head.next;
        size--;
        return value;
    }

    // Devuelve el elemento del principio de la lista sin eliminarlo
    T peek() {
        if (head == null) {
            throw new NoSuchElementException();
        }
        return head.value;
    }

    // Devuelve el número de elementos en la lista
    int size() {
        return size;
    }

    // Invierte el orden de la lista en su lugar
    void reverse() {
        Node<T> prev = null;
        Node<T> current = head;
        Node<T> next;

        while (current != null) {
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        head = prev;
    }

    // Convierte la lista enlazada en un array
    @SuppressWarnings("unchecked")
    T[] asArray(Class<T> cls) {
        List<T> list = new ArrayList<>();
        Node<T> current = head;
        while (current != null) {
            list.add(current.value);
            current = current.next;
        }
        
        T[] array = (T[]) Array.newInstance(cls, list.size());
        for (int i = 0; i < list.size(); i++) {
            array[i] = list.get(i);
        }
        return array;
    }

    // Convierte la lista enlazada en una java.util.List
    List<T> toList() {
        List<T> list = new ArrayList<>();
        Node<T> current = head;
        while (current != null) {
            list.add(current.value);
            current = current.next;
        }
        return list;
    }
}